# Hypothesis Testing

The purpose of the test is to tell if there is any significant difference between two data sets.



## Overview

This module covers,

1) One sample and Two sample t-tests

2) ANOVA

3) Type I and Type II errors

4) Chi-Squared Tests

## Question 1 

*A student is trying to decide between two GPUs. He want to use the GPU for his research to run Deep learning algorithms, so the only thing he is concerned with is speed.*

*He picks a Deep Learning algorithm on a large data set and runs it on both GPUs 15 times, timing each run in hours. Results are given in the below lists GPU1 and GPU2.*

In [2]:
from scipy import stats 
import numpy as np

In [3]:
GPU1 = np.array([11,9,10,11,10,12,9,11,12,9,11,12,9,10,9])
GPU2 = np.array([11,13,10,13,12,9,11,12,12,11,12,12,10,11,13])

#Assumption: Both the datasets (GPU1 & GPU 2) are random, independent, parametric & normally distributed

Hint: You can import ttest function from scipy to perform t tests 

**First T test**

*One sample t-test*

Check if the mean of the GPU1 is equal to zero.
- Null Hypothesis is that mean is equal to zero.
- Alternate hypothesis is that it is not equal to zero.

In [4]:
from scipy.stats             import ttest_1samp,ttest_ind, wilcoxon
from statsmodels.stats.power import ttest_power
import matplotlib.pyplot     as     plt

print('Mean is %2.1f Sd is %2.1f' % (GPU1.mean(),np.std(GPU1,ddof = 1)))

Mean is 10.3 Sd is 1.2


In [7]:
t_statistic, p_value = ttest_1samp(GPU1, 10.3)
print(t_statistic, p_value)

0.10985884360050992 0.9140805576999014


In this example, p value is 0.9140805576999014 and it is more than 5% level of significance

So the statistical decision is to reject the null hypothesis at 5% level of significance.

## Question 2

Given,

Null Hypothesis : There is no significant difference between data sets

Alternate Hypothesis : There is a significant difference

*Do two-sample testing and check whether to reject Null Hypothesis or not.*

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html

In [8]:
t_statistic, p_value  =  stats.ttest_ind(GPU1,GPU2)
print('P Value %1.3f' % p_value)  

P Value 0.014


In this example, p value is 0.014 and it is less than 5% level of significance

So the statistical decision is to reject the null hypothesis at 5% level of significance.

## Question 3

He is trying a third GPU - GPU3.

In [0]:
GPU3 = np.array([9,10,9,11,10,13,12,9,12,12,13,12,13,10,11])

#Assumption: Both the datasets (GPU1 & GPU 3) are random, independent, parametric & normally distributed

*Do two-sample testing and check whether there is significant differene between speeds of two GPUs GPU1 and GPU3.*

#### Answer:

In [9]:
GPU1 = np.array([11,9,10,11,10,12,9,11,12,9,11,12,9,10,9])
GPU3 = np.array([9,10,9,11,10,13,12,9,12,12,13,12,13,10,11])
D  = GPU1 - GPU3
print(D)
print('Mean is %3.2f and standard deviation is %3.2f' %(D.mean(),np.std(D,ddof = 1)))

[ 2 -1  1  0  0 -1 -3  2  0 -3 -2  0 -4  0 -2]
Mean is -0.73 and standard deviation is 1.79


In [10]:
t_statistic, p_value  =  stats.ttest_1samp(D, 0)
print('P Value %1.3f' % p_value)

P Value 0.135


In this example, p value is 0.135 and it is more than 5% level of significance

So the statistical decision is to accept the null hypothesis at 5% level of significance.

## ANOVA

## Question 4 

If you need to compare more than two data sets at a time, an ANOVA is your best bet. 

*The results from three experiments with overlapping 95% confidence intervals are given below, and we want to confirm that the results for all three experiments are not significantly different.*

But before conducting ANOVA, test equality of variances (using Levene's test) is satisfied or not. If not, then mention that we cannot depend on the result of ANOVA

In [12]:
import numpy as np

e1 = np.array([1.595440,1.419730,0.000000,0.000000])
e2 = np.array([1.433800,2.079700,0.892139,2.384740])
e3 = np.array([0.036930,0.938018,0.995956,1.006970])

#Assumption: All the 3 datasets (e1,e2 & e3) are random, independent, parametric & normally distributed

Perform levene test on the data

The Levene test tests the null hypothesis that all input samples are from populations with equal variances. Levene’s test is an alternative to Bartlett’s test bartlett in the case where there are significant deviations from normality.

source: scipy.org

#### Answer:

In [13]:
t_stats, p_value = stats.levene(e1,e2,e3)
print(t_stats, p_value)

2.6741725711150446 0.12259792666001798


In [ ]:
In this example, p value is 0.1226 and it is more than 5% level of significance

So the statistical decision is to accept the null hypothesis at 5% level of significance.
all input samples are from populations with equal variance

## Question 5

The one-way ANOVA tests the null hypothesis that two or more groups have the same population mean. The test is applied to samples from two or more groups, possibly with differing sizes.

use stats.f_oneway() module to perform one-way ANOVA test

In [15]:
import pandas            as pd
import numpy             as np
import matplotlib.pyplot as plt
import seaborn           as sns

In [17]:
e1 = np.array([1.595440,1.419730,0.000000,0.000000])
e2 = np.array([1.433800,2.079700,0.892139,2.384740])
e3 = np.array([0.036930,0.938018,0.995956,1.006970])

print('Count, Mean and standard deviation of population of members of e1: %3d, %3.2f and %3.2f' % (len(e1), e1.mean(),np.std(e1,ddof =1)))
print('Count, Mean and standard deviation of population of members of e2: %3d, %3.2f and %3.2f' % (len(e2), e2.mean(),np.std(e2,ddof =1)))
print('Count, Mean and standard deviation of population of members of e3: %3d, %3.2f and %3.2f' % (len(e3), e3.mean(),np.std(e3,ddof =1)))

Count, Mean and standard deviation of population of members of e1:   4, 0.75 and 0.87
Count, Mean and standard deviation of population of members of e2:   4, 1.70 and 0.67
Count, Mean and standard deviation of population of members of e3:   4, 0.74 and 0.47


In [20]:
population_df = pd.DataFrame()

df1            = pd.DataFrame({'group': '1', 'population':e1})
df2            = pd.DataFrame({'group': '2', 'population':e2})
df3            = pd.DataFrame({'group': '3', 'population':e3})

population_df = population_df.append(df1) 
population_df =  population_df.append(df2) 
population_df =  population_df.append(df3) 

In [23]:
import statsmodels.api         as     sm
from   statsmodels.formula.api import ols
 
mod = ols('population ~ group', data = population_df).fit()
aov_table = sm.stats.anova_lm(mod, typ=2)
print(aov_table)

            sum_sq   df         F    PR(>F)
group     2.399066  2.0  2.513576  0.135746
Residual  4.294994  9.0       NaN       NaN


In [ ]:
In this example, p value is 0.135746 and it is less than our chosen level of signifance at 5%

So the statistical decision is to reject the null hypothesis at 5% level of significance.

## Question 6

*In one or two sentences explain about **TypeI** and **TypeII** errors.*

#### Answer:

1. Type I and Type II errors cannot happen at  the same time 
2. A Type I error can only occur if H0 is true 
3. A Type II error can only occur if H0 is false
4. If Type I error probability (α) increases Type II error probability (β) decreases. 
   Type I & II Error Relationship DCOVA theory.


answer here

## Question 7 

You are a manager of a chinese restaurant. You want to determine whether the waiting time to place an order has changed in the past month from its previous population mean value of 4.5 minutes. 
State the null and alternative hypothesis.

#### Answer:


answer here

1. In testing whether the mean waiting time in the restaurant is more than 4.5 min, the null hypothesis states that mean IQ,  𝜇  equals 100. The alternative hypthesis states that the mean waiting time,  𝜇  is greater than 4.5 min.

𝐻0 :  𝜇  = 4.5
𝐻𝐴 :  𝜇  > 4.5
    
2. Decide the significance level
Here we select  𝛼  = 0.05 and it is given that sample size = n

3. Identify the test statistic
We know the population standard deviation and the sample is a large sample, n>30.
So you use the normal distribution and the  𝑍𝑆𝑇𝐴𝑇  test statistic.

We collect the sample data, calculate the test statistic. In our example,

𝑋 = sample mean
𝜇  = 4.5, population mean
𝜎  = standard deviation
n = sample size
𝑍𝑆𝑇𝐴𝑇 = 𝑋 - 𝜇/(𝜎/√n)

4. In this example, if Z lies in the rejection region because, Z > Zstat, then the statistical decision is to reject the null hypothesis and accept the alternate hypothesis

else we fail to reject the null hypothesis



## Chi square test

## Question 8

Let's create a small dataset for dice rolls of four players

In [24]:
import numpy as np

d1 = [5, 8, 3, 8]
d2 = [9, 6, 8, 5]
d3 = [8, 12, 7, 2]
d4 = [4, 16, 7, 3]
d5 = [3, 9, 6, 5]
d6 = [7, 2, 5, 7]

dice = np.array([d1, d2, d3, d4, d5, d6])

run the test using SciPy Stats library

Depending on the test, we are generally looking for a threshold at either 0.05 or 0.01. Our test is significant (i.e. we reject the null hypothesis) if we get a p-value below our threshold.

For our purposes, we’ll use 0.01 as the threshold.

use stats.chi2_contingency() module 

This function computes the chi-square statistic and p-value for the hypothesis test of independence of the observed frequencies in the contingency table

Print the following:

- chi2 stat
- p-value
- degree of freedom
- contingency



In [32]:
import scipy.stats as stats
import scipy

chi_sq_Stat, p_value, deg_freedom, exp_freq = stats.chi2_contingency(dice)

print('Chi-square statistic %3.5f P value %1.6f Degrees of freedom %d' %(chi_sq_Stat, p_value,deg_freedom))

Chi-square statistic 23.31567 P value 0.077664 Degrees of freedom 15


In [34]:
stats.chi2_contingency(dice)

(23.315671914716496,
 0.07766367301496693,
 15,
 array([[ 5.57419355,  8.20645161,  5.57419355,  4.64516129],
        [ 6.50322581,  9.57419355,  6.50322581,  5.41935484],
        [ 6.73548387,  9.91612903,  6.73548387,  5.61290323],
        [ 6.96774194, 10.25806452,  6.96774194,  5.80645161],
        [ 5.34193548,  7.86451613,  5.34193548,  4.4516129 ],
        [ 4.87741935,  7.18064516,  4.87741935,  4.06451613]]))

in this example p value is 0.077664 is > 0.01 threshold value and hence we accept the null hypothesis

## Question 9

### Z-test

Get zscore on the above dice data using stats.zscore module from scipy. Convert zscore values to p-value and take mean of the array.

In [31]:
Zcrit = round(stats.norm.isf(q = 0.05),2)
print('Value of Z critical is %3.6f' %Zcrit)

Value of Z critical is 1.640000


## Question 10

A Paired sample t-test compares means from the same group at different times.

The basic two sample t-test is designed for testing differences between independent groups. 
In some cases, you might be interested in testing differences between samples of the same group at different points in time. 
We can conduct a paired t-test using the scipy function stats.ttest_rel(). 

In [25]:
before= stats.norm.rvs(scale=30, loc=100, size=500) ## Creates a normal distribution with a mean value of 100 and std of 30
after = before + stats.norm.rvs(scale=5, loc=-1.25, size=500)

Test whether a weight-loss drug works by checking the weights of the same group patients before and after treatment using above data.

In [28]:
import  scipy.stats             as      stats  
t_statistic, p_value  =  stats.ttest_rel(after,before )
print(p_value)  

1.5366732425900215e-06


In this example, p value is 1.5366732425900215e-06 and it is less than 5% level of significance

So the statistical decision is to reject the null hypothesis at 5% level of significance.

So there is sufficient evidence to reject the null hypothesis that there is an effect of drug on weight loss